### PR TITLE
Pop-up bootstrap jquery and other

### PR DESCRIPTION
**Pull request by sonishc**
Pop-up rewrite
Add jquery
Add bootstrap.js
Add validation errors to front-end
### [Show on Heroku](http://cyber-dashboard.herokuapp.com)

For reviewers
- [ ] @afurm 
- [ ] @rubycop 
- [ ] @SetterVet 
- [ ] @VasylYaremkiv

### CODE REVIEW

File: app/javascript/components/users/show/CreateEventPopUp.vue [hunk 1]
```diff
@@ -140,16 +143,22 @@
...
+        if (this.targetDescription) {
+          this.cleanErrorMessages('.erorrs', '.fields-error')
+          this.targets.push({
+            id: null,
+            description: this.targetDescription,
+            achieved: false,
+            _destroy: null
+          });
+          this.targetDescription = '';
+        }else {
```
Comment (afurm) [inline app/javascript/components/users/show/CreateEventPopUp.vue]: spaces

File: app/javascript/components/users/show/CreateEventPopUp.vue [hunk 2]
```diff
@@ -176,8 +185,47 @@
...
           }
-        }).then((response) => location.reload())
-          .catch((error) => console.log(error.response));
+        }).then((response) => this.showNotice(response.data))
+      },
+      showNotice(response) {
+        (() => { 
+          if (response.status == 'created' || response.status == 'updated') {
+            return location.reload()
+          }else {
```
Comment (afurm) [inline app/javascript/components/users/show/CreateEventPopUp.vue]: space

File: app/javascript/components/users/show/CreateEventPopUp.vue [hunk 3]
```diff
@@ -176,8 +185,47 @@
...
+            }
+          }
+        })();
+      },
+      cleanErrorMessages(errorsMessagesClassName, errorsFieldsClassName) {
+        let errors_messages = document.querySelectorAll(errorsMessagesClassName);
+        let errors_fields = document.querySelectorAll(errorsFieldsClassName);
+
+        if (errors_fields) {
+          [].forEach.call(errors_fields, function(el) {
```
Comment (afurm) [inline app/javascript/components/users/show/CreateEventPopUp.vue]: What is the purpose iterating on an empty Array?

File: app/models/event.rb [hunk 4]
```diff
@@ -18,4 +19,17 @@ def data
       status: status
     }
   end
+
+  private
+
+  def date_validation
+    if !self[:finish_date].nil? && !self[:start_date].nil?
```
Comment (afurm) [inline app/models/event.rb]: `if !self[:finish_date].nil? && !self[:start_date].nil?`  -> `return if finish_date.blank? && start_date.blank?`
You do not need use `self`  .

File: app/javascript/components/users/show/CreateEventPopUp.vue (hunk 3 above)
Comment (sonishc) [reply app/javascript/components/users/show/CreateEventPopUp.vue]: > What is the purpose iterating on an empty Array?

It is an alias for Array.prototype.forEach.call(...);
Short example
```
[1, 2, 3].forEach.call(["a", "b", "c"], function (item, i, arr) {
    console.log(i + ": " + item);
});
```
### Result 
0: "a"
1: "b"
2: "c"

File: app/models/event.rb (hunk 4 above)
Comment (sonishc) [reply app/models/event.rb]: > `if !self[:finish_date].nil? && !self[:start_date].nil?` -> `return if finish_date.blank? && start_date.blank?`
> You do not need use `self` .

Validation doesn't work without `self`

File: app/javascript/components/users/show/CreateEventPopUp.vue [hunk 5]
```diff
@@ -176,8 +185,47 @@
...
+          } else {
+            this.cleanErrorMessages('.erorrs', '.fields-error');
+
+            for (let error in response) {
+              this.addErrorMessages(error, response[error], 3)
+            }
+          }
+        })();
+      },
+      cleanErrorMessages(errorsMessagesClassName, errorsFieldsClassName) {
```
Comment (VasylYaremkiv) [inline app/javascript/components/users/show/CreateEventPopUp.vue]: Do we need params for `cleanErrorMessages` function?

File: app/models/event.rb [hunk 6]
```diff
@@ -18,4 +19,17 @@ def data
       status: status
     }
   end
+
+  private
+
+  def date_validation
+    return unless !self[:finish_date].nil? && !self[:start_date].nil?
```
Comment (VasylYaremkiv) [inline app/models/event.rb]: Do not use `self`.
https://chodounsky.net/2014/05/07/when-to-use-self-in-activerecord-models/
`finish_date.present?`

File: app/models/event.rb [hunk 7]
```diff
@@ -18,4 +19,17 @@ def data
...
+
+    if self[:finish_date] < self[:start_date]
+      errors.add(:finish_date, "can't be less than start date")
```
Comment (VasylYaremkiv) [inline app/models/event.rb]: Use localization for error text: https://stackoverflow.com/questions/17935597/ruby-on-rails-i18n-want-to-translate-custom-messages-in-models

File: app/models/event.rb [hunk 8]
```diff
@@ -18,4 +19,17 @@ def data
...
+      false
```
Comment (VasylYaremkiv) [inline app/models/event.rb]: I think we do not need to return any data in validation

File: app/javascript/src/styles/layout/_forms.scss [hunk 9]
```diff
@@ -0,0 +1,26 @@
+.erorrs {
```
Comment (VasylYaremkiv) [inline app/javascript/src/styles/layout/_forms.scss]: `erorrs` block has few errors or  single error?

File: app/javascript/src/styles/layout/_forms.scss [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+  color:red;
+  &::first-letter {
+    text-transform: capitalize;
+  }
+}
+
+.fields-error {
```
Comment (VasylYaremkiv) [inline app/javascript/src/styles/layout/_forms.scss]: Same here

File: app/javascript/components/users/show/CreateEventPopUp.vue [hunk 11]
```diff
@@ -176,8 +185,47 @@
               targets_attributes: this.targets
             }
           }
-        }).then((response) => location.reload())
-          .catch((error) => console.log(error.response));
+        }).then((response) => this.showNotice(response.data))
+      },
+      showNotice(response) {
+        (() => {
```
[the remark this answers is not in the document]
Comment (sonishc) [reply app/javascript/components/users/show/CreateEventPopUp.vue]: > Why do you need inline function?

Ok. Removed. Previously function contains more then one arrow function

File: app/javascript/components/users/UsersList.vue [hunk 12]
```diff
@@ -71,43 +94,58 @@ export default {
...
         let modifier = 1;
-        if (this.currentSortDir === 'desc') {modifier = -1;}
-        if (a[this.currentSort].toLowerCase() < b[this.currentSort].toLowerCase()) return -1 * modifier;
-        if (a[this.currentSort].toLowerCase() > b[this.currentSort].toLowerCase()) return modifier;
+        if (this.currentSortDir === 'desc') {
+          modifier = -1;
+        }
+        if (a[this.currentSort].toLowerCase() < b[this.currentSort].toLowerCase()) {
+          return -1 * modifier;
+        }
```
Comment (sonishc) [inline app/javascript/components/users/UsersList.vue]: @VasylYaremkiv

File: app/javascript/src/styles/layout/_forms.scss [hunk 13]
```diff
@@ -0,0 +1,27 @@
+#popUpModal {
+  .erorr {
```
Comment (afurm) [inline app/javascript/src/styles/layout/_forms.scss]: `.erorr` - > `error`

**Pull request merged**